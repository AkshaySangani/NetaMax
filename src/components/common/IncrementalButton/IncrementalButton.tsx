import React, { ReactElement } from "react";

import { BiMinus, BiPlus } from "react-icons/bi";

import { HStack, IconButton, Input } from "@chakra-ui/react";

import { IIncrementalButtonProps } from "./IIncrementalButtonProps";

/**
 * IncrementalButton component.
 * @param {IIncrementalButtonProps} props the component props.
 * @returns {ReactElement} the component element.
 */
export const IncrementalButton = (props: IIncrementalButtonProps): ReactElement => {
  const {
    min,
    max,
    defaultValue,
    onAdd,
    onSubtract,
    customWidth,
    customHeight,
    customBorderRadius,
  } = props;

  const currentValue = defaultValue && min && defaultValue > min ? defaultValue : min;
  const isDisabled = Boolean(currentValue === max);

  return (
    <HStack
      width={customWidth || "120px"}
      height={customHeight || { base: "27px", md: "32px" }}
      rounded={customBorderRadius || "full"}
      borderWidth="1px"
      overflow="hidden"
    >
      <IconButton
        variant="ghost"
        aria-label="remove"
        icon={<BiMinus />}
        onClick={onSubtract}
        height="100%"
      />
      <Input
        variant="unstyled"
        data-testid="counter"
        readOnly
        textAlign="center"
        value={currentValue}
      />
      <IconButton
        variant="ghost"
        aria-label="add"
        icon={<BiPlus />}
        onClick={onAdd}
        disabled={isDisabled}
        height="100%"
      />
    </HStack>
  );
};
