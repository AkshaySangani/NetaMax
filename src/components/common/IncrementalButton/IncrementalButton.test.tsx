import { render, cleanup, fireEvent } from "@testing-library/react";
import { IncrementalButton } from "./IncrementalButton";
import { IIncrementalButtonProps } from "./IIncrementalButtonProps";

describe("IncrementalButton Component", () => {
  const onAdd = jest.fn();
  const onSubtract = jest.fn();
  let mock: IIncrementalButtonProps;

  beforeEach(() => {
    mock = {
      min: 1,
      max: 10,
      defaultValue: 1,
      width: 115,
      customHeight: "150px",
      customWidth: "200px",
      customBorderRadius: "200px",
      onSubtract,
      onAdd,
    };
  });

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test("IncrementalButton should be render correctly", () => {
    const { container } = render(<IncrementalButton {...mock} />);
    expect(container).toMatchSnapshot();
  });

  test("Add button should working properly", () => {
    const { getByLabelText, getByTestId } = render(<IncrementalButton {...mock} />);
    fireEvent.click(getByLabelText("add"));
    expect(getByTestId("counter")).toHaveValue("1");
    expect(onAdd).toBeCalledTimes(1);
  });

  test("Substract button should working properly", () => {
    const { getByLabelText } = render(<IncrementalButton {...mock} />);
    fireEvent.click(getByLabelText("remove"));
    expect(onSubtract).toBeCalledTimes(1);
  });

  test("Render with default props", () => {
    const { getByLabelText, getByTestId } = render(<IncrementalButton {...mock} />);
    const input = getByTestId("counter");
    expect(input).toHaveValue("1");
    expect(getByLabelText("add")).not.toBeDisabled();
  });

  test("Current value should calculate properly", () => {
    mock = {
      min: 1,
      max: 10,
      defaultValue: 9,
      width: 115,
      customHeight: "150px",
      customWidth: "200px",
      customBorderRadius: "200px",
      onSubtract,
      onAdd,
    };
    const { getByLabelText, getByTestId } = render(<IncrementalButton {...mock} />);
    const input = getByTestId("counter");
    expect(input).toHaveValue("9");
    expect(getByLabelText("add")).not.toBeDisabled();
  });

  test("Add button should be disabled if the value exceed the max value.", () => {
    mock = {
      min: 1,
      max: 10,
      defaultValue: 10,
      width: 115,
      customHeight: "150px",
      customWidth: "200px",
      customBorderRadius: "200px",
      onSubtract,
      onAdd,
    };
    const { getByLabelText, getByTestId } = render(<IncrementalButton {...mock} />);
    const input = getByTestId("counter");
    expect(input).toHaveValue("10");
    expect(getByLabelText("add")).toBeDisabled();
  });
});
