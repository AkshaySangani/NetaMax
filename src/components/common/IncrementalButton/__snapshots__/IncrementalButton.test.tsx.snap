// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncrementalButton Component IncrementalButton should be render correctly 1`] = `
<div>
  <div
    class="chakra-stack css-4qlusk"
  >
    <button
      aria-label="remove"
      class="chakra-button css-ljqi8d"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 11h14v2H5z"
        />
      </svg>
    </button>
    <input
      aria-readonly="true"
      class="chakra-input css-xi606m"
      data-testid="counter"
      readonly=""
      value="1"
    />
    <button
      aria-label="add"
      class="chakra-button css-ljqi8d"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        focusable="false"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 11h-6V5h-2v6H5v2h6v6h2v-6h6z"
        />
      </svg>
    </button>
  </div>
</div>
`;
